### PR TITLE
[SFSRAP-27] Added validation for plugin and repo keys

### PR DESCRIPTION
**Pull request by duncanpMS**
Added helper methods to create valid keys.
Added unit tests.

### CODE REVIEW

File: PluginGenerator/PluginDefinition.cs [hunk 1]
```diff
@@ -28,19 +27,7 @@ public PluginManifest()
...
-            set
-            {
-                this.pluginKey = value;
-                if (this.pluginKey != null)
-                {
-                    this.pluginKey = this.pluginKey.ToLowerInvariant();
-                }
-            }
-        }
+        public string Key { get; set; }
```
Comment (duncanpMS) [inline PluginGenerator/PluginDefinition.cs]: The validation is being performed elsewhere; it doesn't need to be performed here too.

File: PluginGenerator/RulesPluginBuilder.cs [hunk 2]
```diff
@@ -73,6 +74,13 @@ public RulesPluginBuilder SetSqaleFilePath(string filePath)
             return this;
         }
 
+        public RulesPluginBuilder SetRepositoryKey(string key)
+        {
+            RepositoryKeyUtilities.ThrowIfInvalid(key);
+            this.repositoryKey = key;
+            return this;
+        }
+
```
Comment (duncanpMS) [inline PluginGenerator/RulesPluginBuilder.cs]: Previously there wasn't a way to explicitly set the repository key; it was being calculated internally by the rule plugin builder. It now needs to be set explicitly by the caller.

File: PluginGenerator/RulesPluginBuilder.cs [hunk 3]
```diff
@@ -88,7 +96,7 @@ protected override void ApplyConfiguration(string baseWorkingDirectory)
             this.AddRuleSources(tempDir);
             this.AddRuleJars();
 
-            this.SetSourceCodeTokenReplacement(WellKnownSourceCodeTokens.Rule_PluginId, this.FindPluginKey());
```
Comment (duncanpMS) [inline PluginGenerator/RulesPluginBuilder.cs]: _Rule_PluginId_ was a misleading name: it's actually the unique part of the repository id.

File: RoslynPluginGenerator/AnalyzerPluginGenerator.cs [hunk 4]
```diff
@@ -234,11 +234,14 @@ private void BuildPlugin(RoslynPluginDefinition definition, string baseTempDirec
...
                 definition.Manifest.Key + "-plugin-" + definition.Manifest.Version + ".jar");
 
+            string repoKey = RepositoryKeyUtilities.GetValidKey(definition.PackageId + "." + definition.Language);
+
             RulesPluginBuilder builder = new RulesPluginBuilder(logger);
             builder.SetLanguage(definition.Language)
-                            .SetRulesFilePath(definition.RulesFilePath)
-                            .SetProperties(definition.Manifest)
-                            .SetJarFilePath(fullJarPath);
+                        .SetRepositoryKey(repoKey)
```
Comment (duncanpMS) [inline RoslynPluginGenerator/AnalyzerPluginGenerator.cs]: Now need to explicitly set the repository key.

File: Tests/IntegrationTests/Roslyn/RoslynGenTests.cs [hunk 5]
```diff
@@ -59,25 +59,26 @@ public void RoslynPlugin_GenerateForValidAnalyzer_Succeeds()
...
-            AssertExpectedPropertyDefinitionValue("analyzer1.pkgid1.staticResourceName", "Analyzer1.Pkgid1.1.0.2.zip", jarInfo);
-            AssertExpectedPropertyDefinitionValue("analyzer1.pkgid1.pluginKey", "analyzer1.pkgid1", jarInfo);
-            AssertExpectedPropertyDefinitionValue("analyzer1.pkgid1.pluginVersion", "1.0.2", jarInfo);
+            AssertExpectedPropertyDefinitionValue("analyzer1.pkgid1.cs.analyzerId", "Analyzer1.Pkgid1", jarInfo);
+            AssertExpectedPropertyDefinitionValue("analyzer1.pkgid1.cs.ruleNamespace", "Analyzer1.Pkgid1", jarInfo);
+            AssertExpectedPropertyDefinitionValue("analyzer1.pkgid1.cs.nuget.packageId", "Analyzer1.Pkgid1", jarInfo);
+            AssertExpectedPropertyDefinitionValue("analyzer1.pkgid1.cs.nuget.packageVersion", "1.0.2", jarInfo);
+            AssertExpectedPropertyDefinitionValue("analyzer1.pkgid1.cs.staticResourceName", "Analyzer1.Pkgid1.1.0.2.zip", jarInfo);
+            AssertExpectedPropertyDefinitionValue("analyzer1.pkgid1.cs.pluginKey", "analyzer1pkgid1", jarInfo);
+            AssertExpectedPropertyDefinitionValue("analyzer1.pkgid1.cs.pluginVersion", "1.0.2", jarInfo);
```
Comment (duncanpMS) [inline Tests/IntegrationTests/Roslyn/RoslynGenTests.cs]: I've amended the repository key to include the language as a prefix. This means would could put VB and C# repositories/analyser assemblies in the same jar in the future if we wanted to.

File: Tests/IntegrationTests/Roslyn/RoslynGenTests.cs [hunk 6]
```diff
@@ -59,25 +59,26 @@ public void RoslynPlugin_GenerateForValidAnalyzer_Succeeds()
 
             Assert.IsNotNull(jarInfo, "Failed to process the generated jar successfully");
 
-            AssertExpectedManifestValue(WellKnownPluginProperties.Key, "analyzer1.pkgid1", jarInfo);
+            AssertExpectedManifestValue(WellKnownPluginProperties.Key, "analyzer1pkgid1", jarInfo);
```
Comment (duncanpMS) [inline Tests/IntegrationTests/Roslyn/RoslynGenTests.cs]: The plugin key can't contain "." characters now.

**Comment by bgavrilMS** [comment]
LGTM

**Pull request merged**